doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+
+  * Incrementally define all necessary functionality through the new API (NEP 42):
+
+    * Defining operations such as ``np.common_type``.
+    * Allowing to define casting between datatypes.
+    * Add functionality necessary to create a numpy array from Python scalars
+      (i.e. ``np.array(...)``).
+    * …
+
+  * Restructure how universal functions functions to (NEP 43):
+
+    * make it possible to allow a `~numpy.ufunc` such as ``np.add`` to be
+      extended by user defined datatypes such as Units.
+
+    * allow efficient lookup for the correct implementation for user defined
+      datatypes.
+
+    * enable reuse of existing code. Units should be able to use the
+      normal math loops and add additional logic to determine output type.
+
+* Phase III: Growth of NumPy and Scientific Python Ecosystem capabilities:
+
+  * Cleanup of legacy behaviour where it is considered buggy or undesirable.
+  * Provide a path to define new datatypes from Python.
+  * Assist the community in creating types such as Units or Categoricals
+  * Allow strings to be used in functions such as ``np.equal`` or ``np.add``.
+  * Remove legacy code paths within NumPy to improve long term maintainability
+
+This document serves as a basis for phase I and provides the vision and
+motivation for the full project.
+While of limited use to the end user, the first step involves a conceptional
+cleanup of the current datatype system. It provides a more "pythonic" datatype
+Python type object, with a clear class hierarchy.
+
+The second phase is the incremental creation of all APIs necessary to define
+fully featured datatypes and reorganization of the NumPy datatype system.
+Phase I concerns mainly private APIs, while phase II is expected to include
+a slow creation of a *preliminary* public API which will develop into the
+final stable API.
+
+Some of the benefits of a large refactor may only become evident after the full
+deprecation of the current legacy implementation (i.e. larger code removals).
+However, these steps are necessary for improvements to many parts of the
+core NumPy API, and are expected to make the implementation generally
+easier to understand.
+Pushing forward with Phase I cleans up the concepts of the API to coincide
+with how Python is typically organized and unlocks the ability to improve
+parts of NumPy which currently are not extensible.
+
+The following picture gives another overview of the overall design idea.
+Note that this NEP only regards Phase I (shaded area),
+the rest encompasses Phase II and the design choices are up for discussion,
+however, it highlights that the DType datatype class is the central, necessary
+concept:
+
+.. image:: _static/nep-0041-mindmap.svg
+
+
+First steps directly related to this NEP
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The required changes necessary to NumPy are massive and touch a large part
+of the code base.
+We thus propose the implementation of the above steps listed in Phases I and II.
```

Comment:
This sentence is a bit unclear. The point that I got from these two sentences was the fact that the changes will be so extensive is what persuaded you to take a multi-stage approach.